Newsletters: add 282 (2023 year in review) (#1436)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2023-12-20-newsletter.md

ability to perform all operations manually using pen, paper,
instructions, and paper cutouts, including generating an encoded seed
(with dice), protecting it with a checksum, creating checksummed shares,
verifying checksums, and recovering the seed. This allows users to
verify share integrity individually on a regular basis without relying
on trusted computing devices.

## March

{:#mpchan}
In March, pseudonymonous developer John Law [published][mar mpchan] a
paper describing a way to create a hierarchy of channels for multiple
users from a single onchain transaction.  The design can allow
all online users to spend their funds even when some of their channel
counterparties are offline, which is not currently possible in LN.  This
optimization would allow always-online users to use their capital more
efficiently, likely reducing costs of other users of LN.  The proposal
depends on Law's tunable penalty protocol, which has not received any
public software development since it was proposed in 2022.

![Tunable Penalty Protocol](/img/posts/2023-03-tunable-commitment.dot.png)

<div markdown="1" class="callout" id="softforks">
### 2023 summary<br>Soft fork proposals

> A draft BIP for the `OP_VAULT` opcode was [posted][feb op_vault] in

Author: Good catch, thanks!

single-hop, single-direction atomic transfer protocol from the owner to
the counterparty for various uses such as mixing coins, making internal
transfers, and paying LN invoices.  Concerns were raised about the high
onchain footprint, large hot wallet requirements, and capital costs
associated with this protocol compared to LN.  However, several
developers remained enthusiastic about the proposed protocol and its
potential to provide a simple and trustless experience for its users.

## June

{:#silentpayments}
Josie Baker and Ruben Somsen [posted][jun sp] a draft BIP for [silent
payments][topic silent payments], a type of reusable payment code that
will produce a unique onchain address each time it is used, preventing
[output linking][topic output linking].  Output linking can
significantly reduce the privacy of users (including users not directly
involved in a transaction). The draft goes into detail about the
benefits of the proposal, its tradeoffs, and how software can
effectively use it.

<div markdown="1" class="callout" id="security">
### 2023 summary<br>Security disclosures

Optech reported on three significant security vulnerabilities this year:

> - [Milksad vulnerability in Libbitcoin `mx`][aug milksad]: a widely

Reviewer: Wasn't it bx?

```suggestion
- [Milksad vulnerability in Libbitcoin `bx`][aug milksad]: a widely
```

Reviewer 2: It was indeed: https://github.com/bitcoinbook/bitcoinbook/blob/develop/appdx-bx.asciidoc

keys. The beta release supports CLN and LDK, layer-1 and layer-2
validation rules, backup and recovery capabilities, and provides a
reference implementation.

{:#ln-meeting}
An LN developer [meeting][jul summit] held in July discussed a variety
of topics, including reliable transaction confirmation at the base
layer, [taproot][topic taproot] and [MuSig2][topic musig] channels,
updated channel announcements, [PTLCs][topic ptlc] and [redundant
overpayment][topic redundant overpayments], [channel jamming mitigation
proposals][topic channel jamming attacks], simplified commitments, and
the specification process.

## August

{:#onion-messages}
[Support][aug onion] for [onion messages][topic onion messages] was
added to the LN specification in August.  Onion messages allow sending
one-way messages across the network. Like payments ([HTLCs][topic
htlc]), the messages use onion encryption so that each forwarding node
only knows what peer it received the message from and what peer should
next receive the message. Message payloads are also encrypted so that
only the ultimate receiver can read it.  Onion messages use [blinded

> paths][topic rv routing], which was added to the LN specification in

Reviewer 3: ```suggestion
paths][topic rv routing], which were added to the LN specification in
```

[P2TR][topic taproot] with support for [MuSig2][topic musig]-style
[scriptless multisignature signing][topic multisignature] when both
parties are cooperating.  This reduces transaction weight and
improves privacy when channels are closed cooperatively. LND continues
to exclusively use [HTLCs][topic htlc], allowing payments starting in a
taproot channel to continue to be forwarded through other LN nodes that
don’t support taproot channels.

## September

{:#compressed-txes}
In September, Tom Briar [published][sept compress] a draft specification
and implementation of compressed Bitcoin transactions. The proposal
addresses the challenge of compressing uniformly distributed data in
Bitcoin transactions by replacing integer representations with
variable-length integers, using block height and location to reference
transactions instead of their outpoint txid, and omitting public keys
from P2WPKH transactions. While the compressed format saves space,
converting it back into a usable format requires more CPU, memory, and
I/O than processing regular serialized transactions, which is an
acceptable tradeoff in situations such as broadcast by satellite or

> stegnographic transfer.

Reviewer 3: ```suggestion
steganographic transfer.
```

<div markdown="1" class="callout" id="releases">
### 2023 summary<br>Major releases of popular infrastructure projects

- [Eclair 0.8.0][jan eclair] added support for [zero-conf
  channels][topic zero-conf channels] and Short Channel IDentifier

>   (SCID) aliases

Reviewer 3: ```suggestion
  (SCID) aliases.
```

- [HWI 2.2.0][jan hwi] added support for [P2TR][topic taproot] keypath
  spends using the BitBox02 hardware signing device.

- [Core Lightning 23.02][mar cln] added experimental support for peer
  storage of backup data and updates experimental support for [dual
  funding][topic dual funding] and [offers][topic offers].

- [Rust Bitcoin 0.30.0][mar rb] provided a large number of API changes,
  along with the announcement of a new website.

- [LND v0.16.0-beta][apr lnd] provided a new major version of this
  popular LN implementation.

- [Libsecp256k1 0.3.1][apr secp] fixed an issue related to code that

>   should have run in constant time but was not when compiled with Clang

Reviewer 3: ```suggestion
  should have run in constant time but did not when compiled with Clang
```

  version 14 or higher.

- [LDK 0.0.115][apr ldk] included more support for the experimental
  [offers protocol][topic offers] and improved security and privacy.

- [Core Lightning 23.05][may cln] is a release of the newest version of
  this LN implementation. It includes support for [blinded
  payments][topic rv routing], version 2 [PSBTs][topic psbt], and more

>   flexible feerate management

Reviewer 3: ```suggestion
  flexible feerate management.
```

- [Bitcoin Core 25.0][may bcc] added a new `scanblocks` RPC, simplified
  the use of `bitcoin-cli`, added [miniscript][topic miniscript] support
  to the `finalizepsbt` RPC, reduced default memory use with the
  `blocksonly` configuration option, and sped up wallet rescans when

>   [compact block filters][topic compact block filters] are enabled

Reviewer 3: ```suggestion
  [compact block filters][topic compact block filters] are enabled.
```

- [LND v0.17.0-beta][oct lnd] included experimental support for “simple
  taproot channels”, which allows using [unannounced channels][topic
  unannounced channels] funded onchain using a [P2TR][topic taproot]
  output. This is the first step towards adding other features to LND’s
  channels, such as support for Taproot Assets and [PTLCs][topic ptlc].
  The release also includes a significant performance improvement for
  users of the Neutrino backend, which supports [compact block
  filters][topic compact block filters], as well as improvements to
  LND’s built-in [watchtower][topic watchtowers] functionality.

- [LDK 0.0.117][oct ldk] included security bug fixes related to the
  [anchor outputs][topic anchor outputs] features included in the
  immediately prior release. The release also improved pathfinding,
  improved [watchtower][topic watchtowers] support, and enabled
  [batch][topic payment batching] funding of new channels.

- [LDK 0.0.118][nov ldk] included partial experimental support for the
  [offers protocol][topic offers].

- [Core Lightning 23.11][nov cln] provided additional flexibility to the
  _rune_ authentication mechanism, improved backup verification, new and

>   features for plugins.

Reviewer 3: There seems to be something missing from this sentence.

and Carol each start with 1 BTC, the state can be updated so that Alice
has 2 BTC, Bob 0 BTC, and Carol 1 BTC.  Participants would still use and
advertise regular LN channels, and if those channels became imbalanced,
a rebalancing can be done through an offchain peerswap within the state
contract. This method is private to participants, requires less onchain
space, and likely eliminates offchain rebalancing fees, thus improving
revenue potential for forwarding nodes and reliability for LN payments.
However, it requires a multiparty state contract, which is untested in
production.  ZmnSCPxj suggests using [LN-Symmetry][topic eltoo] or
duplex payment channels as a foundation, with duplex channels being more
feasible despite their inability to remain open indefinitely.

{:#assumeutxo}
October saw the [completion][oct assumeutxo] of the first phase of the
[assumeUTXO project][topic assumeutxo], containing all the remaining
changes necessary to both use an assumedvalid snapshot chainstate and do
a full validation sync in the background. It makes UTXO snapshots
loadable via an RPC.  Although the feature set will not be available on

> mainnet until activated, this merge marked the culmination of a

Reviewer: "Activated" makes it sound like it were a softfork, rather than it being unavailable due to no mainnet UTXO set snapshot being shipped with the latest release

> mainnet until activated, this merge marked the culmination of a

Reviewer 4: Suggest alternate wording for "activated" here. Maybe something about the AssumeutxoHash / check being in the code in a future release?

multi-year effort. The project, proposed in 2018 and formalized in 2019,
will significantly improve the user experience of new full nodes first
coming onto the network.

{:#p2pv2}
Also [accomplished][oct p2pv2] by the Bitcoin Core project in October was the
addition of support for [version 2 encrypted P2P transport][topic v2 p2p
transport] as specified in [BIP324][]. The feature is currently disabled
by default but can be enabled using the `-v2transport` option.

> Encrypted transport helps improve the privacy of Bitcoin users by
> preventing passive observers (such as ISPs) from directly determining
> which transactions nodes relay to their peers. It’s also possible to use

Reviewer: ```suggestion
Encrypted transport improves the privacy of Bitcoin users by
preventing passive observers (such as ISPs) from observing
which transactions nodes relay to their peers. While previously a man-in-the-middle attacker would have been able to simply observe or censor relay of certain transactions, either reading or modifying traffic would now require an active observer that maintains the full protocol state with both Bitcoin nodes whose traffic is being observed. It’s also possible to use
```

Author: I think this is too much detail given that we link to our earlier post.  Also, "maintaining the full protocol state" for BIP324 isn't a big deal.  I don't see how it'd be significantly more complicated than performing a pure TCP MITM.  As an example, there's plenty of tools for MITM'ing SSL connections.

> encrypted transport to thwart active man-in-the-middle observers by

Reviewer 3: ```suggestion
encrypted transport to detect active man-in-the-middle observers by
```

comparing session identifiers. In the future, the addition of [other
features][topic countersign] may make it convenient for a lightweight
client to securely connect to a trusted node over a P2P encrypted
connection.

{:#miniscript}

> Several additional improvements to Bitcoin Core were made through the
> year in the form of miniscript descriptor support.  February saw the
> [ability][feb miniscript] to create miniscript descriptors for P2WSH
> output scripts.  October saw miniscript support [updated][oct
> miniscript] to support taproot, including miniscript descriptors for
> tapscript.

Reviewer: ```suggestion
Miniscript descriptor support in Bitcoin Core saw several improvements through the
year.  February saw the
[ability][feb miniscript] to create miniscript descriptors for P2WSH
output scripts.  October saw miniscript support [updated][oct
miniscript] to support taproot, including miniscript descriptors for
tapscript.
```

{:#statebitvm}
A method for state compression in Bitcoin using zero-knowledge validity
proofs was [described][may state] by Robin Linus and Lukas George in
May.  This massively reduces the amount of state that a client needs to
download in order to trustlessly verify future operations in a system,
for example starting a new full node with only a relatively small
validity proof rather than validating every already confirmed
transaction on the block chain.  In October, Robin Linus
[introduced][oct bitvm] BitVM, a method enabling bitcoins to be paid
contingent on the successful execution of an arbitrary program, without
requiring a consensus change in Bitcoin.  BitVM requires substantial
offchain data exchange but needs only a single onchain transaction for
agreement, or a small number of onchain transactions if there's a

> dispute. BitVM makes possible complex trustless contracts even in
> adversarial scenarios, which has caught the attention of several
> developers.

Reviewer: ```suggestion
dispute. BitVM makes complex trustless contracts possible even in
adversarial scenarios, which has caught the attention of several
developers.
```

## November

{:#offers}
With the final specification of [blinded paths][topic rv routing] and
[onion messages][topic onion messages], and their implementation in
multiple popular LN nodes, a significant amount of progress was made
through this year in development of the [offers protocol][topic offers]

> that depended on them.  Offers allow a receiver's wallet to generate a

Reviewer: ```suggestion
that depends on them.  Offers allow a receiver's wallet to generate a
```

published over 86,000 English words about Bitcoin software research and
development this year, the rough equivalent of a 250-page book.

In addition, every newsletter this year was accompanied by a [podcast][]
episode, totaling over 50 hours in audio form and 450,000 words in
transcript form.  Many of Bitcoin's top contributors were guests on the
show---some of them on more than one episode---with a total of 63
different unique guests in 2023:

- Abubakar Ismail
- Adam Gibson
- Alekos Filini
- Alex Myers
- Anthony Towns
- Antoine Poinsot
- Armin Sabouri
- Bastien Teinturier
- Brandon Black
- Burak Keceli
- Calvin Kim
- Carla Kirk-Cohen
- Christian Decker
- Clara Shikhelman
- Dan Gould
- Dave Harding
- Dusty Daemon
- Eduardo Quintana
- Ethan Heilman
- Fabian Jahr
- Gijs van Dam
- Gloria Zhao
- Gregory Sanders
- Henrik Skogstrøm
- Jack Ronaldi
- James O’Beirne
- Jesse Posner
- Johan Torås Halseth
- Joost Jager
- Josie Baker
- Ken Sedgwick
- Larry Ruane
- Lisa Neigut
- Lukas George
- Martin Zumsande
- Matt Corallo
- Matthew Zipkin
- Max Edwards
- Maxim Orlovsky
- Nick Farrow
- Niklas Gögge
- Olaoluwa Osuntokun
- Pavlenex
- Peter Todd
- Pieter Wuille

> - Rearden Code

Reviewer: Rearden Code is Brandon Black
```suggestion
```

Author: Thanks!

- Robin Linus
- Ruben Somsen
- Russell O’Connor
- Salvatore Ingala
- Sergi Delgado Segura
- Severin Bühler
- Steve Lee
- Steven Roose
- Thomas Hartman
- Thomas Voegtlin
- Tom Briar
- Tom Trevethan
- Valentine Wallace
- Vivek Kasarabada
- Will Clark
- Yuval Kogman
- ZmnSCPxj

Optech also published two contributed field reports from the business
community: one from Brandon Black at BitGo about [implementing
MuSig2][] for reducing fee costs and improving privacy, and a second
report from Antoine Poinsot at Wizardsardine about [building software
with miniscript][].

[implementing musig2]: /en/bitgo-musig2/
[building software with miniscript]: /en/wizardsardine-miniscript/

</div>

## December

{:#clustermempool}

> Several Bitcoin Core developers [began][may cluster] working on a design to

Reviewer 4: Might put the cluster mempool name in here directly.

simplify mempool operations while maintaining the necessary transaction
ordering, where parent transactions must be confirmed before their
children.  Transactions are grouped into clusters, then split into
feerate-sorted chunks, ensuring that high feerate chunks are confirmed
first.  This allows creation of block templates by simply choosing the
highest feerate chunks in the mempool, and eviction of transactions by
dropping the lowest feerate chunks.  This fixes some existing
undesirable behavior (e.g., where miners might lose out on fee revenue
due to suboptimal evictions) and may be able to improve other aspects of
mempool management and transaction relay in the future.  Archives of
their discussions were [published][dec cluster] in early December.

{:#warnet}
December also saw the public [announcement][dec warnet announce] of new
tool for launching a large number of Bitcoin nodes with a defined set of
connections between them (usually on a test network).  This can be used

> to test behavior that's difficult to replicate using small numbers nodes

Reviewer 4: ```suggestion
to test behavior that's difficult to replicate using small numbers of nodes
```

[aug opreturn]: /en/newsletters/2023/08/09/#proposed-changes-to-bitcoin-core-default-relay-policy
[aug payjoin]: /en/newsletters/2023/08/16/#serverless-payjoin
[chatbtc]: https://chat.bitcoinsearch.xyz/
[dec bcc]: /en/newsletters/2023/12/06/#bitcoin-core-26-0
[dec cluster]: /en/newsletters/2023/12/06/#cluster-mempool-discussion
[dec liqad]: /en/newsletters/2023/12/13/#discussion-about-griefing-liquidity-ads
[dec payjoin]: /en/newsletters/2023/12/13/#payjoin-client-for-bitcoin-core-released
[dec warnet announce]: /en/newsletters/2023/12/13/#bitcoin-network-simulation-tool-warnet-announced
[dec zipkin warnet]: /en/newsletters/2023/12/06/#testing-with-warnet
[feb bitcoinsearch]: /en/newsletters/2023/02/15/#bitcoinsearch-xyz
[feb cln offers]: /en/newsletters/2023/02/08/#core-lightning-5892
[feb codex32]: /en/newsletters/2023/02/22/#proposed-bip-for-codex32-seed-encoding-scheme
[feb eclair offers]: /en/newsletters/2023/02/22/#eclair-2479
[feb htlcendo]: /en/newsletters/2023/02/22/#feedback-requested-on-ln-good-neighbor-scoring
[feb lnflag]: /en/newsletters/2023/02/22/#ln-quality-of-service-flag
[feb miniscript]: /en/newsletters/2023/02/22/#bitcoin-core-24149

> [feb op_vault2]: /en/newsletters/2023/03/08/#alternative-design-for-op-vault

Reviewer: This topic was described in a March newsletter and the source email is from March 1st.

Author: Yeah, but I think we described it in the February section.  The month names in these are just to reduce the chance of a accidentally redefinition of a previously used link.

[feb op_vault]: /en/newsletters/2023/02/22/#draft-bip-for-op-vault
[feb payjoin]: /en/newsletters/2023/02/01/#serverless-payjoin-proposal
[feb storage]: /en/newsletters/2023/02/15/#core-lightning-5361
[jamming paper]: /en/newsletters/2022/11/16/#paper-about-channel-jamming-attacks
[jan bip329]: /en/newsletters/2023/01/25/#bips-1383
[jan eclair]: /en/newsletters/2023/01/04/#eclair-0-8-0
[jan hwi]: /en/newsletters/2023/01/18/#hwi-2-2-0
[jan inquisition]: /en/newsletters/2023/01/04/#bitcoin-inquisition
[jan ord1]: /en/newsletters/2023/02/08/#discussion-about-storing-data-in-the-block-chain

> [jan ord2]: /en/newsletters/2023/02/15/#continued-discussion-about-block-chain-data-storage

Reviewer: I just noticed that these appear in February newsletters. The first topic links to an email from January, but the second links to an email from February.

co-author Clara Shikhelman began [asking][feb htlcendo] for feedback on
the suggested parameters to use when implementing one of the ideas from
the paper, [HTLC endorsement][topic htlc endorsement].  In April, they
[posted][apr htlcendo] a draft specification for their testing plans.
In July, the idea and the proposal were [discussed][jul htlcendo] at the
LN developers meeting, which led to some mailing list discussion about
an [alternative approach][aug antidos] to make the costs paid by both
attackers and honest users reflect the underlying costs paid by the node
operators providing the service; that way, a node operator who is
earning a reasonable return on providing services to honest users will
continue earning reasonable returns if attackers begin using those
services.  In August, it was [announced][aug htlcendo] developers
associated with Eclair, Core Lightning, and LND were all implementing
parts of the HTLC endorsement protocol in order to begin collecting data
related to it.

{:#codex32}
Russell O'Connor and Andrew Poelstra [proposed][feb codex32] a new BIP
for backing up and restoring [BIP32][] seeds called [codex32][topic

> codex32].  Similar to SLIP39, it allows creating several shares using

Author: What's the problem with "allows creating"?  I'm not a big fan of "one" as a pronoun.

Shamir’s Secret Sharing Scheme with configurable threshold requirements.
An attacker who obtains less than the threshold number of shares will
learn nothing about the seed. Unlike other recovery codes that use a
wordlist, codex32 uses the same alphabet as [bech32][topic bech32]
addresses. The main advantage of codex32 over existing schemes is its
ability to perform all operations manually using pen, paper,
instructions, and paper cutouts, including generating an encoded seed
(with dice), protecting it with a checksum, creating checksummed shares,
verifying checksums, and recovering the seed. This allows users to
verify share integrity individually on a regular basis without relying
on trusted computing devices.

## March

{:#mpchan}

> In March, pseudonymonous developer John Law [published][mar mpchan] a

Author: Oof, good catch, thanks!  I gotta add pseudonymous to my spell checker so I don't skip past misspelled versions of it.

paper describing a way to create a hierarchy of channels for multiple
users from a single onchain transaction.  The design can allow
all online users to spend their funds even when some of their channel
counterparties are offline, which is not currently possible in LN.  This
optimization would allow always-online users to use their capital more

> efficiently, likely reducing costs of other users of LN.  The proposal

Reviewer 2: ```suggestion
efficiently, likely reducing the costs of other users of LN.  The proposal
```

<div markdown="1" class="callout" id="softforks">
### 2023 summary<br>Soft fork proposals

A [proposal][jan op_vault] for a new `OP_VAULT` opcode was posted in
January by James O'Beirne, [followed][feb op_vault] in February by a
draft BIP an implementation for Bitcoin Inquisition.  This was further
followed a few weeks later by a [proposal][feb op_vault2] for an
alternative design for `OP_VAULT` by Gregory Sanders.

The merklize all the things (MATT) proposal first described last year saw
activity again this year, starting with Salvatore Ingala [showing][may
matt] how it could provide most of the features of the proposed
`OP_VAULT` opcodes.  Johan Torås Halseth [further demonstrated][jun
matt] how one of the opcodes from the MATT proposal could replicate
key functions of the `OP_CHECKTEMPLATEVERIFY` proposed opcode, although the
MATT version was less space efficient.  Halseth also used the
opportunity to introduce readers to a tool he'd built, [tapsim][], which
allows debugging Bitcoin transactions and [tapscripts][topic tapscript].

In June, Robin Linus [described][jun specsf] how users could timelock funds today,

> use them on a sidechain for an extended period of time, and then allow

Reviewer 2: ```suggestion
use them on a sidechain for an extended time, and then allow
```

factories without user interaction, reducing the risk of last-minute
onchain congestion and transaction fees. Anthony Towns raised concerns
about the _forced expiration flood_ problem, where the failure of a
large user could force many time-sensitive onchain transactions
simultaneously. Law replied to note that he is working on a solution to
delay expiry during high transaction fee periods.

October started with Steven Roose [posting][oct txhash] a draft BIP for
a new OP_TXHASH opcode. The idea for the opcode has been discussed
before but this is the first specification of the idea. In addition to
describing exactly how the opcode would work, it examined some
downsides, such as full nodes potentially needing to hash up to several
megabytes of data every time the opcode is invoked. The BIP draft
included a sample implementation of the opcode.

Also in October, Rusty Russell [researched][oct generic] generic
covenants with minimal changes to Bitcoin's scripting language and Ethan
Heilman [published][oct op_cat] a draft BIP to add an [OP_CAT][op_cat]
opcode that would concatenate two elements on the stack.  The discussion

> for both of those topics would [continue][nov cov] on into November.

Reviewer 2: ```suggestion
for both of those topics would [continue][nov cov] into November.
```

efficient onchain and could make it more difficult to perform certain
types of [channel jamming attacks][topic channel jamming attacks].

</div>
## April

{:#watchaccount}
Sergi Delgado Segura [proposed][apr watchtower] an accountability
mechanism for [watchtowers][topic watchtowers] for cases where they fail to respond to
protocol breaches that they were capable of detecting. For example,
Alice provides a watchtower with data for detecting and responding to
the confirmation of an old LN channel state; later, that state is
confirmed but the watchtower fails to respond. Alice would like to be
able to hold the watchtower operator accountable by publicly proving it
failed to respond appropriately.  Delgado suggests a mechanism based on
cryptographic accumulators that watchtowers can use for creating
commitments, which users can later use to produce accountability proofs
if a breach occurs.

{:#route-blinding}
[Route blinding][topic rv routing], first described three years earlier,
was [added][apr blinding] to the LN specification in April.  It allows a
receiver to provide a spender with the identifier of a particular
forwarding node and an onion-encrypted path from that node to the

> receiver's own node.  The spender forwards a payment and the encrypted

Reviewer 2: ```suggestion
receiver's node.  The spender forwards a payment and the encrypted
```

proposed protocol for [splicing][topic splicing], which allows nodes to
continue using a channel while funds are added to it or removed from it.
This is especially useful for keeping funds in a channel while still
allowing instant onchain payments to be made from that balance, allowing
a wallet user interface to show users a single balance from which they
can make either onchain or offchain payments.  By the end of the year,
both Core Lightning and Eclair would support splicing.

![Splicing](/img/posts/2023-04-splicing1.dot.png)

## May

{:#lspspec}
A set of draft specifications for Lightning Service Providers (LSPs) was
[released][may lsp] in May.  Standards make it easy for a client to
connect to multiple LSPs, which will prevent vendor lock-in and improve
privacy.  The first specification released describes an API for allowing
a client to purchase a channel from an LSP, achieving a functionality
similar to [liquidity advertisements][topic liquidity advertisements].
The second describes an API for setting up and managing [just-in-time
(JIT) channels][topic jit channels].

{:#payjoin}
Dan Gould spent a significant part of the year working on enhancements

> to the [payjoin][topic payjoin] protocol, a privacy enhancing technique

Reviewer 2: ```suggestion
to the [payjoin][topic payjoin] protocol, a privacy-enhancing technique
```

that makes it much more difficult for a third party to reliably
associate inputs and outputs in a transaction with either the spender or
receiver.  In February, he [proposed][feb payjoin] a serverless payjoin

> protocol than can be used even if the receiver doesn't operate an

Reviewer 2: ```suggestion
protocol that can be used even if the receiver doesn't operate an
```

before the timelock expires. The protocol provides a trustless
single-hop, single-direction atomic transfer protocol from the owner to
the counterparty for various uses such as mixing coins, making internal
transfers, and paying LN invoices.  Concerns were raised about the high
onchain footprint and the need for the operator to keep a large amount
of funds in a hot wallet compared to LN.  However, several
developers remained enthusiastic about the proposed protocol and its
potential to provide a simple and trustless experience for its users.

## June

{:#silentpayments}
Josie Baker and Ruben Somsen [posted][jun sp] a draft BIP for [silent
payments][topic silent payments], a type of reusable payment code that
will produce a unique onchain address each time it is used, preventing
[output linking][topic output linking].  Output linking can
significantly reduce the privacy of users (including users not directly
involved in a transaction). The draft goes into detail about the
benefits of the proposal, its tradeoffs, and how software can
effectively use it.  Ongoing work implementing silent payments for
Bitcoin Core was also [discussed][aug sp] during a Bitcoin Core PR

> Review Club meeeting.

Reviewer 2: ```suggestion
Review Club meeting.
```

<div markdown="1" class="callout" id="security">
### 2023 summary<br>Security disclosures

Optech reported on three significant security vulnerabilities this year:

- [Milk Sad vulnerability in Libbitcoin `bx`][aug milksad]: a widely
  undocumented lack of entropy in a command suggested for creating

>   wallets ultimately led the theft of a significant number of

Reviewer 2: ```suggestion
  wallets ultimately led to the theft of a significant number of
```

  unaware of the vulnerability being exploited.

- [Replacement cycling against HTLCs][oct cycling]: a fund-stealing
  attack against [HTLCs][topic htlc] used in LN and possibly other
  protocols was privately discovered and responsibly disclosed by
  Antoine Riard.  All LN implementations tracked by Optech deployed
  mitigations, although the effectiveness of those mitigations was a
  subject of discussion and other mitigations have been proposed.

</div>

## July

{:#vls}
The Validating Lightning Signer (VLS) project [released][jul vls] their
first beta version in July.  The project allows the separation of an LN
node from the keys that control its funds. An LN node running with VLS
will route signing requests to a remote signing device instead of local
keys. The beta release supports CLN and LDK, layer-1 and layer-2
validation rules, backup and recovery capabilities, and provides a
reference implementation.

{:#ln-meeting}
An LN developer [meeting][jul summit] held in July discussed a variety
of topics, including reliable transaction confirmation at the base
layer, [taproot][topic taproot] and [MuSig2][topic musig] channels,

> updated channel announcements, [PTLCs][topic ptlc] and [redundant

Reviewer 4: Could reference another (September) PTLC topic here, but a bit tortured perhaps:

- LN messaging changes for PTLCs - https://bitcoinops.org/en/newsletters/2023/09/13/#ln-messaging-changes-for-ptlcs

Author: Yeah, I thought about that but I got the impression (perhaps mistakenly) that people didn't find that discussion useful, so I decided not to draw attention to it.  (Insert here the meme "looks-at-you-looks-away")

Reviewer 4: I think its fine to leave out

overpayment][topic redundant overpayments], [channel jamming mitigation
proposals][topic channel jamming attacks], simplified commitments, and

> the specification process.

Reviewer 4: Two other LN related July items could be appended here as well with probably not much trauma to the main structure of the YiR:

- LN spec cleanup https://bitcoinops.org/en/newsletters/2023/07/12/#ln-specification-clean-up-proposed 
- Simplified LN closing protocol https://bitcoinops.org/en/newsletters/2023/07/26/#simplified-ln-closing-protocol

> the specification process.

Reviewer 2: I wonder if this would be better as a list, especially because of the compound items "taproot and MuSig2 channels" (very different from "taproot, and MuSig2 channels"), and "PTLCs and redundant overpayment"-- one must look at the commas and the "and"s very carefully.

Author: Keeping as-is.  The concern is real, but my goal is to keep each item to a single paragraph (since this is already a long summary) and anyone who wants details can read our original summary which does have each item in a separate list (with more details about each).

signed a higher version number.  This mechanism isn't Bitcoin-specific,
but incorporating certain Bitcoin opcodes could enable its use onchain.
In a Lightning Network (LN) channel, that would allow Alice to claim all
channel funds if Bob provided an outdated backup, thus reducing the risk
of Bob deceiving Alice and stealing her balance.  The proposal sparked
significant discussion.  Peter Todd pointed out its versatility beyond
LN and suggested a simpler mechanism without fraud proofs, while Ghost43
highlighted the importance of such proofs when dealing with anonymous
peers.

{:#tapchan}
LND added [experimental support][aug lnd taproot] for "simple taproot
channels", allowing LN funding and commitment transactions to use
[P2TR][topic taproot] with support for [MuSig2][topic musig]-style
[scriptless multisignature signing][topic multisignature] when both
parties are cooperating.  This reduces transaction weight and
improves privacy when channels are closed cooperatively. LND continues
to exclusively use [HTLCs][topic htlc], allowing payments starting in a
taproot channel to continue to be forwarded through other LN nodes that
don’t support taproot channels.

> 

Reviewer 4: There are 2 related-in-my-mind items that could fall under a new August heading around "policy":

- January's Continued RBF discussion https://bitcoinops.org/en/newsletters/2023/01/04/#continued-rbf-discussion
- Aug's Proposed changes to Bitcoin Core default relay policy  https://bitcoinops.org/en/newsletters/2023/08/09/#proposed-changes-to-bitcoin-core-default-relay-policy 

a 3th, weaker related item:
- July's Preventing coinjoin pinning https://bitcoinops.org/en/newsletters/2023/06/28/#preventing-coinjoin-pinning-with-v3-transaction-relay

This could be another opportunity to highlight the waiting for confirmation series (even though we do already call it out) which wrapped up end of July.

Just a thread I noticed in my notes of things not covered in this YiR while going through the 2023 newsletters.

Author: We already cover the Aug proposed relay changes item, although we associate it with its op_return side in the ordinals paragraph.

For the RBF discussion, we had a whole extended section on that in last year's YIR, and the january item was really just the tail end of that (IMO), so I think it's ok that we skip it.

The coinjoin/v3 item doesn't feel strong enough to me to warrant a parargaph of its own, or even a paragraph with just it and the mempool series (since we already give that a short paragraph).  Hopefully we'll have a lot of v3 stuff next year.

Thanks for going through each of the newsletters again!  I meant to do that myself but got lazy.

{:#compressed-txes}
In September, Tom Briar [published][sept compress] a draft specification
and implementation of compressed Bitcoin transactions. The proposal
addresses the challenge of compressing uniformly distributed data in
Bitcoin transactions by replacing integer representations with
variable-length integers, using block height and location to reference
transactions instead of their outpoint txid, and omitting public keys
from P2WPKH transactions. While the compressed format saves space,
converting it back into a usable format requires more CPU, memory, and
I/O than processing regular serialized transactions, which is an
acceptable tradeoff in situations such as broadcast by satellite or
steganographic transfer.

<div markdown="1" class="callout" id="releases">
### 2023 summary<br>Major releases of popular infrastructure projects

- [Eclair 0.8.0][jan eclair] added support for [zero-conf
  channels][topic zero-conf channels] and Short Channel IDentifier
  (SCID) aliases.

- [HWI 2.2.0][jan hwi] added support for [P2TR][topic taproot] keypath
  spends using the BitBox02 hardware signing device.

- [Core Lightning 23.02][mar cln] added experimental support for peer

>   storage of backup data and updates experimental support for [dual

Reviewer 2: ```suggestion
  storage of backup data and updated experimental support for [dual
```

- [Rust Bitcoin 0.30.0][mar rb] provided a large number of API changes,
  along with the announcement of a new website.

- [LND v0.16.0-beta][apr lnd] provided a new major version of this
  popular LN implementation.

- [Libsecp256k1 0.3.1][apr secp] fixed an issue related to code that
  should have run in constant time but did not when compiled with Clang
  version 14 or higher.

- [LDK 0.0.115][apr ldk] included more support for the experimental
  [offers protocol][topic offers] and improved security and privacy.

- [Core Lightning 23.05][may cln] included support for [blinded
  payments][topic rv routing], version 2 [PSBTs][topic psbt], and more
  flexible feerate management.

- [Bitcoin Core 25.0][may bcc] added a new `scanblocks` RPC, simplified
  the use of `bitcoin-cli`, added [miniscript][topic miniscript] support
  to the `finalizepsbt` RPC, reduced default memory use with the
  `blocksonly` configuration option, and sped up wallet rescans when
  [compact block filters][topic compact block filters] are enabled.

- [Eclair v0.9.0][jun eclair] was a release that "contains a lot of
  preparatory work for important (and complex) lightning features:

>   [dual-funding][topic dual funding], [splicing][topic splicing] and

Reviewer 2: ```suggestion
  [dual-funding][topic dual funding], [splicing][topic splicing], and
```

  BOLT12 [offers][topic offers]."

- [HWI 2.3.0][jul hwi] added support for DIY Jade devices and a binary
  for running the main hwi program on Apple Silicon hardware with MacOS
  12.0+.

- [LDK 0.0.116][jul ldk] included support for [anchor outputs][topic
  anchor outputs] and [multipath payments][topic multipath payments]
  with [keysend][topic spontaneous payments].

- [BTCPay Server 1.11.x][aug btcpay] included improvements to invoice
  reporting, additional upgrades to the checkout process, and new

>   features for the point of sale terminal.

Reviewer 2: ```suggestion
  features for the point-of-sale terminal.
```

attackers using PSS.  Additionally, PSS offers increased Lightning
Network throughput and may help mitigate [channel jamming attacks][topic
channel jamming attacks].

{:#sidepools}
Developer ZmnSCPxj [proposed][oct sidepool] a concept called _sidepools_
that aims to enhance LN's liquidity management. Sidepools involve
multiple forwarding nodes depositing funds into a multiparty offchain
state contract similar to LN channels. This allows funds to be
redistributed among participants offchain. For instance, if Alice, Bob,
and Carol each start with 1 BTC, the state can be updated so that Alice
has 2 BTC, Bob 0 BTC, and Carol 1 BTC.  Participants would still use and
advertise regular LN channels, and if those channels became imbalanced,
a rebalancing can be done through an offchain peerswap within the state
contract. This method is private to participants, requires less onchain
space, and likely eliminates offchain rebalancing fees, thus improving
revenue potential for forwarding nodes and reliability for LN payments.
However, it requires a multiparty state contract, which is untested in
production.  ZmnSCPxj suggests building on [LN-Symmetry][topic eltoo]

> (which requires a consensus change) or duplex payment channels (whach

Reviewer 2: ```suggestion
(which requires a consensus change) or duplex payment channels (which
```

loadable via an RPC.  Although the feature set is not yet directly
usable by inexperienced users, this merge marked the culmination of a
multi-year effort. The project, proposed in 2018 and formalized in 2019,
will significantly improve the user experience of new full nodes first
coming onto the network.

{:#p2pv2}
Also [accomplished][oct p2pv2] by the Bitcoin Core project in October was the
addition of support for [version 2 encrypted P2P transport][topic v2 p2p
transport] as specified in [BIP324][]. The feature is currently disabled
by default but can be enabled using the `-v2transport` option.
Encrypted transport helps improve the privacy of Bitcoin users by
preventing passive observers (such as ISPs) from directly determining
which transactions nodes relay to their peers. It’s also possible to use
encrypted transport to detect active man-in-the-middle observers by
comparing session identifiers. In the future, the addition of [other
features][topic countersign] may make it convenient for a lightweight
client to securely connect to a trusted node over a P2P encrypted
connection.

{:#miniscript}
Miniscript descriptor support saw several additional improvements in

> Bitcoin Core through the year.  February saw the

Reviewer 2: ```suggestion
Bitcoin Core throughout the year.  February saw the
```

[ability][feb miniscript] to create miniscript descriptors for P2WSH
output scripts.  October saw miniscript support [updated][oct
miniscript] to support taproot, including miniscript descriptors for
tapscript.

{:#statebitvm}
A method for state compression in Bitcoin using zero-knowledge validity
proofs was [described][may state] by Robin Linus and Lukas George in
May.  This massively reduces the amount of state that a client needs to

> download in order to trustlessly verify future operations in a system,

Reviewer 2: ```suggestion
download to trustlessly verify future operations in a system,
```

for example starting a new full node with only a relatively small
validity proof rather than validating every already confirmed
transaction on the block chain.  In October, Robin Linus
[introduced][oct bitvm] BitVM, a method enabling bitcoins to be paid
contingent on the successful execution of an arbitrary program, without
requiring a consensus change in Bitcoin.  BitVM requires substantial
offchain data exchange but needs only a single onchain transaction for
agreement, or a small number of onchain transactions if there's a
dispute. BitVM makes complex trustless contracts possible even in
adversarial scenarios, which has caught the attention of several
developers.

## November

{:#offers}
With the final specification of [blinded paths][topic rv routing] and
[onion messages][topic onion messages], and their implementation in
multiple popular LN nodes, a significant amount of progress was made

> through this year in development of the [offers protocol][topic offers]

Reviewer 2: ```suggestion
through this year in the development of the [offers protocol][topic offers]
```

that depends on them.  Offers allow a receiver's wallet to generate a
short _offer_ that can be shared with a spender's wallet.  The spender's
wallet can use the offer to contact the receiver's wallet over the LN
protocol to request a specific invoice, which it can then pay in the
usual way.  This allows the creation of reusable offers that can each
produce a different invoice, invoices that can be updated with current
information (e.g. the exchange rate) just seconds before they're paid,
and offers that can be paid by the same wallet more than once (e.g. a
subscription), among other features.  Existing experimental
implementations of offers in [Core Lightning][feb cln offers] and
[Eclair][feb eclair offers] were updated during the year, and support
for offers was added to [LDK][sept ldk offers].  Additionally, November

> saw [discussion][nov offers] about creating an updated version of

Reviewer 2: ```suggestion
saw a [discussion][nov offers] about creating an updated version of
```

- Niklas Gögge
- Olaoluwa Osuntokun
- Pavlenex
- Peter Todd
- Pieter Wuille
- Robin Linus
- Ruben Somsen
- Russell O’Connor
- Salvatore Ingala
- Sergi Delgado Segura
- Severin Bühler
- Steve Lee
- Steven Roose
- Thomas Hartman
- Thomas Voegtlin
- Tom Briar
- Tom Trevethan
- Valentine Wallace
- Vivek Kasarabada
- Will Clark
- Yuval Kogman
- ZmnSCPxj

Optech also published two contributed field reports from the business
community: one from Brandon Black at BitGo about [implementing
MuSig2][] for reducing fee costs and improving privacy, and a second
report from Antoine Poinsot at Wizardsardine about [building software
with miniscript][].

[implementing musig2]: /en/bitgo-musig2/
[building software with miniscript]: /en/wizardsardine-miniscript/

</div>

## December

{:#clustermempool}
Several Bitcoin Core developers [began][may cluster] working on a new
[cluster mempool][topic cluster mempool] design to
simplify mempool operations while maintaining the necessary transaction
ordering, where parent transactions must be confirmed before their
children.  Transactions are grouped into clusters, then split into
feerate-sorted chunks, ensuring that high feerate chunks are confirmed

> first.  This allows creation of block templates by simply choosing the

Reviewer 2: ```suggestion
first.  This allows the creation of block templates by simply choosing the
```

highest feerate chunks in the mempool, and eviction of transactions by
dropping the lowest feerate chunks.  This fixes some existing
undesirable behavior (e.g., where miners might lose out on fee revenue
due to suboptimal evictions) and may be able to improve other aspects of
mempool management and transaction relay in the future.  Archives of
their discussions were [published][dec cluster] in early December.

{:#warnet}

> December also saw the public [announcement][dec warnet announce] of new

Reviewer 2: ```suggestion
December also saw the public [announcement][dec warnet announce] of a new
```

space, and likely eliminates offchain rebalancing fees, thus improving
revenue potential for forwarding nodes and reliability for LN payments.
However, it requires a multiparty state contract, which is untested in
production.  ZmnSCPxj suggests building on [LN-Symmetry][topic eltoo]
(which requires a consensus change) or duplex payment channels (whach
can be implemented today), with duplex channels being more feasible

> today despite their inability to remain open indefinitely.

Reviewer: "(which can be implemented today)" and "being more feasible today" feels a bit like a repetition
